File: template/netwrix-java17/function/src/main/java/com/openfaas/function/Handler.java
package com.accessanalyzer.function;

import com.accessanalyzer.model.IHandler;
import com.accessanalyzer.model.IResponse;
import com.accessanalyzer.model.IRequest;
import com.accessanalyzer.model.Response;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.JsonNode;

import java.util.*;
import java.util.concurrent.TimeUnit;
import java.time.Instant;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.io.*;
import java.nio.file.*;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.URI;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class Handler extends com.accessanalyzer.model.AbstractHandler {

    private static final ObjectMapper objectMapper = new ObjectMapper();
    private static final String SECRETS_DIR = "/var/openfaas/secrets/";
    
    public IResponse Handle(IRequest req) {
        try {
            // Create context from environment variables
            Context context = new Context();
            context.setHostname(System.getenv().getOrDefault("HOSTNAME", "localhost"));
            context.setScanId(System.getenv("SCAN_ID"));
            context.setSyncId(System.getenv("SYNC_ID"));
            context.setRunLocal(System.getenv().getOrDefault("RUN_LOCAL", "false"));
            context.setFunctionType(System.getenv("FUNCTION_TYPE"));
            
            // Parse CONFIG
            String configStr = System.getenv("CONFIG");
            if (configStr != null && !configStr.isEmpty()) {
                try {
                    JsonNode configJson = objectMapper.readTree(configStr);
                    context.setConfig(configJson);
                } catch (Exception e) {
                    return createErrorResponse(400, "Invalid JSON in CONFIG: " + e.getMessage());
                }
            }

            boolean localRun = "true".equals(context.getRunLocal());

            if (localRun) {
                if (context.getConfig() == null) {
                    return createErrorResponse(400, "CONFIG is required when RUN_LOCAL is true");
                }

                if (context.getFunctionType() == null || context.getFunctionType().isEmpty()) {
                    return createErrorResponse(400, "FUNCTION_TYPE is required when RUN_LOCAL is true");
                }

                // Validate request body against config
                String requestBody = req.getBody();
                if (requestBody != null && !requestBody.isEmpty()) {
                    try {
                        JsonNode requestData = objectMapper.readTree(requestBody);
                        ValidationResult validation = LocalTesting.validateRequestSchema(
                            context.getConfig(), requestData, context.getFunctionType());
                        if (!validation.isValid()) {
                            return createErrorResponse(400, validation.getErrorMessage());
                        }
                    } catch (Exception e) {
                        return createErrorResponse(400, "Invalid JSON in request body: " + e.getMessage());
                    }
                }
            }

            // Load secrets from OpenFaaS secret files
            context.setSecrets(getSecrets(localRun));

            // Parse execution IDs from request body
            String requestBody = req.getBody();
            if (requestBody != null && !requestBody.isEmpty()) {
                try {
                    JsonNode requestData = objectMapper.readTree(requestBody);
                    if (requestData.has("scanExecutionId")) {
                        context.setScanExecutionId(requestData.get("scanExecutionId").asText());
                    }
                    if (requestData.has("syncExecutionId")) {
                        context.setSyncExecutionId(requestData.get("syncExecutionId").asText());
                    }
                } catch (Exception e) {
                    // Ignore JSON parsing errors for execution IDs
                }
            }

            if (context.getSecrets().isEmpty()) {
                System.out.println("Warning: No secrets loaded from secret files");
            } else {
                System.out.println("Loaded " + context.getSecrets().size() + " secrets from secret files");
            }

            String startedAt = Instant.now().atOffset(ZoneOffset.UTC).format(DateTimeFormatter.ISO_INSTANT);
            IResponse responseData = handleFunction(req, context);
            String completedAt = Instant.now().atOffset(ZoneOffset.UTC).format(DateTimeFormatter.ISO_INSTANT);

            System.out.println("Response data: " + responseData);

            // Add timestamps for successful responses
            if ("test-connection".equals(context.getFunctionType()) && responseData.getStatusCode() == 200) {
                addTimestampsToResponse(responseData, startedAt, completedAt);
            } else if ("access-scan".equals(context.getFunctionType()) && responseData.getStatusCode() == 200) {
                addTimestampsToResponse(responseData, startedAt, completedAt);
            }

            if (localRun) {
                ValidationResult validation = LocalTesting.validateResponse(context.getFunctionType(), responseData);
                if (!validation.isValid()) {
                    responseData = context.errorResponse(false, validation.getErrorMessage());
                }
            }

            return responseData;

        } catch (Exception e) {
            e.printStackTrace();
            return createErrorResponse(500, "Internal server error: " + e.getMessage());
        }
    }

    private IResponse createErrorResponse(int statusCode, String errorMessage) {
        Response response = new Response();
        response.setStatusCode(statusCode);
        Map<String, Object> body = new HashMap<>();
        body.put("error", errorMessage);
        response.setBody(objectMapper.valueToTree(body).toString());
        return response;
    }

    private void addTimestampsToResponse(IResponse response, String startedAt, String completedAt) {
        try {
            String bodyStr = response.getBody();
            JsonNode bodyJson;
            if (bodyStr == null || bodyStr.isEmpty()) {
                bodyJson = objectMapper.createObjectNode();
            } else {
                bodyJson = objectMapper.readTree(bodyStr);
            }
            
            if (bodyJson.isObject()) {
                ((com.fasterxml.jackson.databind.node.ObjectNode) bodyJson).put("startedAt", startedAt);
                ((com.fasterxml.jackson.databind.node.ObjectNode) bodyJson).put("completedAt", completedAt);
                response.setBody(bodyJson.toString());
            }
        } catch (Exception e) {
            System.err.println("Error adding timestamps to response: " + e.getMessage());
        }
    }

    private Map<String, String> getSecrets(boolean localRun) {
        Map<String, String> secrets = new HashMap<>();
        
        try {
            Path secretsPath = Paths.get(SECRETS_DIR);
            if (!Files.exists(secretsPath)) {
                System.out.println("Secrets directory does not exist: " + SECRETS_DIR);
                return secrets;
            }

            Files.list(secretsPath)
                .filter(Files::isRegularFile)
                .forEach(file -> {
                    String filename = file.getFileName().toString();
                    
                    String keyName;
                    if (localRun) {
                        // For local run, use the filename as-is (no scan ID removal)
                        keyName = filename;
                    } else {
                        // For non-local run, remove the last 9 characters (dash + 8 chars scan ID)
                        if (filename.length() > 9 && filename.charAt(filename.length() - 9) == '-') {
                            keyName = filename.substring(0, filename.length() - 9);
                        } else {
                            System.out.println("Skipping secret file with unexpected format: " + filename);
                            return;
                        }
                    }

                    // Convert dash-separated to camelCase
                    String[] keyParts = keyName.split("-");
                    String camelKey;
                    if (keyParts.length > 1) {
                        camelKey = keyParts[0];
                        for (int i = 1; i < keyParts.length; i++) {
                            camelKey += keyParts[i].substring(0, 1).toUpperCase() + keyParts[i].substring(1);
                        }
                    } else {
                        camelKey = keyParts[0];
                    }

                    // Read the secret content
                    try {
                        String content = Files.readString(file).trim();
                        secrets.put(camelKey, content);
                        System.out.println("Loaded secret: " + camelKey);
                    } catch (IOException e) {
                        System.err.println("Error reading secret file " + filename + ": " + e.getMessage());
                    }
                });

        } catch (IOException e) {
            System.err.println("Error reading secrets directory: " + e.getMessage());
        }

        return secrets;
    }

    // Placeholder for the actual function implementation
    private IResponse handleFunction(IRequest req, Context context) {
        Response res = new Response();
        res.setStatusCode(200);
        Map<String, Object> body = new HashMap<>();
        body.put("message", "Hello from Netwrix Java handler");
        res.setBody(objectMapper.valueToTree(body).toString());
        return res;
    }

    // Context class to hold request context information
    public static class Context {
        private String hostname;
        private Map<String, String> secrets = new HashMap<>();
        private String scanId;
        private String syncId;
        private String scanExecutionId;
        private String syncExecutionId;
        private String runLocal;
        private JsonNode config;
        private String functionType;

        // Getters and setters
        public String getHostname() { return hostname; }
        public void setHostname(String hostname) { this.hostname = hostname; }
        
        public Map<String, String> getSecrets() { return secrets; }
        public void setSecrets(Map<String, String> secrets) { this.secrets = secrets; }
        
        public String getScanId() { return scanId; }
        public void setScanId(String scanId) { this.scanId = scanId; }
        
        public String getSyncId() { return syncId; }
        public void setSyncId(String syncId) { this.syncId = syncId; }
        
        public String getScanExecutionId() { return scanExecutionId; }
        public void setScanExecutionId(String scanExecutionId) { this.scanExecutionId = scanExecutionId; }
        
        public String getSyncExecutionId() { return syncExecutionId; }
        public void setSyncExecutionId(String syncExecutionId) { this.syncExecutionId = syncExecutionId; }
        
        public String getRunLocal() { return runLocal; }
        public void setRunLocal(String runLocal) { this.runLocal = runLocal; }
        
        public JsonNode getConfig() { return config; }
        public void setConfig(JsonNode config) { this.config = config; }
        
        public String getFunctionType() { return functionType; }
        public void setFunctionType(String functionType) { this.functionType = functionType; }

        public IResponse testConnectionSuccessResponse() {
            Response response = new Response();
            response.setStatusCode(200);
            response.setBody("{}");
            return response;
        }

        public IResponse accessScanSuccessResponse() {
            Response response = new Response();
            response.setStatusCode(200);
            response.setBody("{}");
            return response;
        }

        public IResponse getObjectSuccessResponse(byte[] data) {
            Response response = new Response();
            response.setStatusCode(200);
            String encodedData = Base64.getEncoder().encodeToString(data);
            Map<String, Object> body = new HashMap<>();
            body.put("data", encodedData);
            try {
                response.setBody(objectMapper.writeValueAsString(body));
            } catch (Exception e) {
                response.setBody("{\"data\":\"" + encodedData + "\"}");
            }
            return response;
        }

        public IResponse errorResponse(boolean clientError, String errorMsg) {
            Response response = new Response();
            response.setStatusCode(clientError ? 400 : 500);
            Map<String, Object> body = new HashMap<>();
            body.put("error", errorMsg);
            try {
                response.setBody(objectMapper.writeValueAsString(body));
            } catch (Exception e) {
                response.setBody("{\"error\":\"" + errorMsg.replace("\"", "\\\"") + "\"}");
            }
            return response;
        }

        public SaveDataResult saveData(List<Map<String, Object>> data) {
            // Add scan_id, scan_execution_id, and scanned_at to each row
            List<Map<String, Object>> enhancedData = new ArrayList<>();
            String currentTime = Instant.now().atOffset(ZoneOffset.UTC).format(DateTimeFormatter.ISO_INSTANT);

            boolean localRun = "true".equals(this.runLocal);
            String scanId = localRun ? "scan0001" : this.scanId;
            String scanExecutionId = localRun ? "scan-0002" : this.scanExecutionId;

            for (Map<String, Object> row : data) {
                Map<String, Object> enhancedRow = new HashMap<>();
                enhancedRow.put("scan_id", scanId);
                enhancedRow.put("scan_execution_id", scanExecutionId);
                enhancedRow.put("scanned_at", currentTime);
                enhancedRow.putAll(row);
                enhancedData.add(enhancedRow);
            }

            // Dev environment validation
            if (localRun) {
                ValidationResult validation = LocalTesting.validateDevData(this.config, enhancedData);
                if (!validation.isValid()) {
                    System.out.println(validation.getErrorMessage());
                    return new SaveDataResult(false, validation.getErrorMessage());
                } else {
                    System.out.println("Saving " + enhancedData.size() + " items to table");
                    if (!enhancedData.isEmpty()) {
                        try {
                            String sampleJson = objectMapper.writerWithDefaultPrettyPrinter()
                                .writeValueAsString(enhancedData.get(0));
                            System.out.println("Sample item: " + sampleJson);
                        } catch (Exception e) {
                            System.out.println("Sample item: " + enhancedData.get(0));
                        }
                    }
                    return new SaveDataResult(true, null);
                }
            } else {
                String saveDataFunction = System.getenv("SAVE_DATA_FUNCTION");
                if (saveDataFunction == null || saveDataFunction.isEmpty()) {
                    String errorMsg = "SAVE_DATA_FUNCTION is not in the environment";
                    System.out.println(errorMsg);
                    return new SaveDataResult(false, errorMsg);
                }

                Map<String, Object> payload = new HashMap<>();
                payload.put("sourceType", System.getenv("SOURCE_TYPE"));
                payload.put("version", System.getenv("SOURCE_VERSION"));
                payload.put("table", "access");
                payload.put("data", enhancedData);

                try {
                    String payloadJson = objectMapper.writeValueAsString(payload);
                    String url = System.getenv("OPENFAAS_GATEWAY") + "/async-function/" + saveDataFunction;

                    HttpClient client = HttpClient.newBuilder()
                        .connectTimeout(java.time.Duration.ofSeconds(30))
                        .build();

                    HttpRequest request = HttpRequest.newBuilder()
                        .uri(URI.create(url))
                        .header("Content-Type", "application/json")
                        .POST(HttpRequest.BodyPublishers.ofString(payloadJson))
                        .timeout(java.time.Duration.ofSeconds(30))
                        .build();

                    HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());

                    if (response.statusCode() == 202) {
                        return new SaveDataResult(true, null);
                    } else {
                        String errorMsg = "Status " + response.statusCode() + ": " + response.body();
                        System.out.println(errorMsg);
                        return new SaveDataResult(false, errorMsg);
                    }
                } catch (Exception e) {
                    String errorMsg = "Error: " + e.getMessage();
                    System.out.println(errorMsg);
                    return new SaveDataResult(false, errorMsg);
                }
            }
        }

        public UpdateExecutionResult updateExecution(String status, Integer totalObjects, 
                Integer completedObjects, Integer incrementCompletedObjects, String completedAt) {
            // Validation for dev environment
            if ("true".equals(this.runLocal)) {
                ValidationResult validation = LocalTesting.validateUpdateExecutionParams(
                    status, totalObjects, completedObjects, incrementCompletedObjects, completedAt);
                if (!validation.isValid()) {
                    System.out.println(validation.getErrorMessage());
                    return new UpdateExecutionResult(false, validation.getErrorMessage());
                } else {
                    return new UpdateExecutionResult(true, null);
                }
            } else {
                String appUpdateFunction = System.getenv("APP_UPDATE_EXECUTION_FUNCTION");
                if (appUpdateFunction == null || appUpdateFunction.isEmpty()) {
                    String errorMsg = "APP_UPDATE_EXECUTION_FUNCTION is not in the environment";
                    System.out.println(errorMsg);
                    return new UpdateExecutionResult(false, errorMsg);
                }

                String executionId;
                String executionType;

                if (this.scanExecutionId != null && !this.scanExecutionId.isEmpty()) {
                    executionId = this.scanExecutionId;
                    executionType = "scan";
                } else if (this.syncExecutionId != null && !this.syncExecutionId.isEmpty()) {
                    executionId = this.syncExecutionId;
                    executionType = "sync";
                } else {
                    String errorMsg = "Missing required field: either 'scanExecutionId' or 'syncExecutionId' must be provided";
                    System.out.println(errorMsg);
                    return new UpdateExecutionResult(false, errorMsg);
                }

                Map<String, Object> payload = new HashMap<>();
                payload.put("type", executionType);
                payload.put("executionId", executionId);

                // Only include optional fields if they are provided (not null)
                if (status != null) payload.put("status", status);
                if (totalObjects != null) payload.put("totalObjects", totalObjects);
                if (completedObjects != null) payload.put("completedObjects", completedObjects);
                if (incrementCompletedObjects != null) payload.put("incrementCompletedObjects", incrementCompletedObjects);
                if (completedAt != null) payload.put("completedAt", completedAt);

                try {
                    String payloadJson = objectMapper.writeValueAsString(payload);
                    String url = System.getenv("OPENFAAS_GATEWAY") + "/async-function/" + appUpdateFunction;

                    HttpClient client = HttpClient.newBuilder()
                        .connectTimeout(java.time.Duration.ofSeconds(30))
                        .build();

                    HttpRequest request = HttpRequest.newBuilder()
                        .uri(URI.create(url))
                        .header("Content-Type", "application/json")
                        .POST(HttpRequest.BodyPublishers.ofString(payloadJson))
                        .timeout(java.time.Duration.ofSeconds(30))
                        .build();

                    HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());

                    if (response.statusCode() == 202) {
                        return new UpdateExecutionResult(true, null);
                    } else {
                        String errorMsg = "Status " + response.statusCode() + ": " + response.body();
                        System.out.println(errorMsg);
                        return new UpdateExecutionResult(false, errorMsg);
                    }
                } catch (Exception e) {
                    String errorMsg = "Error: " + e.getMessage();
                    System.out.println(errorMsg);
                    return new UpdateExecutionResult(false, errorMsg);
                }
            }
        }
    }

    // Result classes
    public static class ValidationResult {
        private final boolean valid;
        private final String errorMessage;

        public ValidationResult(boolean valid, String errorMessage) {
            this.valid = valid;
            this.errorMessage = errorMessage;
        }

        public boolean isValid() { return valid; }
        public String getErrorMessage() { return errorMessage; }
    }

    public static class SaveDataResult {
        private final boolean success;
        private final String errorMessage;

        public SaveDataResult(boolean success, String errorMessage) {
            this.success = success;
            this.errorMessage = errorMessage;
        }

        public boolean isSuccess() { return success; }
        public String getErrorMessage() { return errorMessage; }
    }

    public static class UpdateExecutionResult {
        private final boolean success;
        private final String errorMessage;

        public UpdateExecutionResult(boolean success, String errorMessage) {
            this.success = success;
            this.errorMessage = errorMessage;
        }

        public boolean isSuccess() { return success; }
        public String getErrorMessage() { return errorMessage; }
    }
}

File: template/netwrix-golang/function/handler.go
package function

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"
)

type Event struct {
	Body    []byte              `json:"body"`
	Headers map[string][]string `json:"headers"`
	Method  string              `json:"method"`
	Query   map[string][]string `json:"query"`
	Path    string              `json:"path"`
}

type Context struct {
	Hostname          string                 `json:"hostname"`
	Secrets           map[string]string      `json:"secrets"`
	ScanID            string                 `json:"scan_id"`
	SyncID            string                 `json:"sync_id"`
	ScanExecutionID   string                 `json:"scan_execution_id"`
	SyncExecutionID   string                 `json:"sync_execution_id"`
	RunLocal          string                 `json:"run_local"`
	Config            map[string]interface{} `json:"config"`
	FunctionType      string                 `json:"function_type"`
}

type ResponseBody struct {
	Error       string      `json:"error,omitempty"`
	Data        string      `json:"data,omitempty"`
	StartedAt   string      `json:"startedAt,omitempty"`
	CompletedAt string      `json:"completedAt,omitempty"`
}

type Response struct {
	StatusCode int                    `json:"statusCode"`
	Body       map[string]interface{} `json:"body"`
	Headers    map[string]string      `json:"headers,omitempty"`
}

func (c *Context) TestConnectionSuccessResponse() Response {
	return Response{
		StatusCode: 200,
		Body:       map[string]interface{}{},
	}
}

func (c *Context) AccessScanSuccessResponse() Response {
	return Response{
		StatusCode: 200,
		Body:       map[string]interface{}{},
	}
}

func (c *Context) GetObjectSuccessResponse(data []byte) Response {
	encodedData := base64.StdEncoding.EncodeToString(data)
	return Response{
		StatusCode: 200,
		Body: map[string]interface{}{
			"data": encodedData,
		},
	}
}

func (c *Context) ErrorResponse(clientError bool, errorMsg string) Response {
	statusCode := 500
	if clientError {
		statusCode = 400
	}

	return Response{
		StatusCode: statusCode,
		Body: map[string]interface{}{
			"error": errorMsg,
		},
	}
}

func (c *Context) SaveData(data []interface{}) (bool, string) {
	// Add scan_id, scan_execution_id, and scanned_at to each row
	var enhancedData []map[string]interface{}
	currentTime := time.Now().UTC().Format(time.RFC3339)

	localRun := c.RunLocal == "true"
	scanID := "scan0001"
	scanExecutionID := "scan-0002"
	
	if !localRun {
		scanID = c.ScanID
		scanExecutionID = c.ScanExecutionID
	}

	for _, row := range data {
		if rowMap, ok := row.(map[string]interface{}); ok {
			enhancedRow := map[string]interface{}{
				"scan_id":           scanID,
				"scan_execution_id": scanExecutionID,
				"scanned_at":        currentTime,
			}
			// Merge original row data
			for k, v := range rowMap {
				enhancedRow[k] = v
			}
			enhancedData = append(enhancedData, enhancedRow)
		}
	}

	// Dev environment validation
	if localRun {
		isValid, errorMsg := ValidateDevData(c.Config, enhancedData)
		if !isValid {
			log.Println(errorMsg)
			return false, errorMsg
		} else {
			log.Printf("Saving %d items to table\n", len(enhancedData))
			if len(enhancedData) > 0 {
				sampleJSON, _ := json.MarshalIndent(enhancedData[0], "", "  ")
				log.Printf("Sample item: %s\n", string(sampleJSON))
			}
			return true, ""
		}
	} else {
		saveDataFunction := os.Getenv("SAVE_DATA_FUNCTION")
		if saveDataFunction == "" {
			errorMsg := "SAVE_DATA_FUNCTION is not in the environment"
			log.Println(errorMsg)
			return false, errorMsg
		}

		payload := map[string]interface{}{
			"sourceType": os.Getenv("SOURCE_TYPE"),
			"version":    os.Getenv("SOURCE_VERSION"),
			"table":      "access",
			"data":       enhancedData,
		}

		payloadJSON, err := json.Marshal(payload)
		if err != nil {
			errorMsg := fmt.Sprintf("Error marshaling payload: %v", err)
			log.Println(errorMsg)
			return false, errorMsg
		}

		url := fmt.Sprintf("%s/async-function/%s", os.Getenv("OPENFAAS_GATEWAY"), saveDataFunction)
		req, err := http.NewRequest("POST", url, bytes.NewBuffer(payloadJSON))
		if err != nil {
			errorMsg := fmt.Sprintf("Error creating request: %v", err)
			log.Println(errorMsg)
			return false, errorMsg
		}

		req.Header.Set("Content-Type", "application/json")
		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()
		req = req.WithContext(ctx)

		client := &http.Client{}
		resp, err := client.Do(req)
		if err != nil {
			errorMsg := fmt.Sprintf("Error: %v", err)
			log.Println(errorMsg)
			return false, errorMsg
		}
		defer resp.Body.Close()

		if resp.StatusCode == 202 {
			return true, ""
		} else {
			body, _ := io.ReadAll(resp.Body)
			errorMsg := fmt.Sprintf("Status %d: %s", resp.StatusCode, string(body))
			log.Println(errorMsg)
			return false, errorMsg
		}
	}
}

func (c *Context) UpdateExecution(status *string, totalObjects *int, completedObjects *int, incrementCompletedObjects *int, completedAt *string) (bool, string) {
	// Validation for dev environment
	if c.RunLocal == "true" {
		isValid, errorMsg := ValidateUpdateExecutionParams(status, totalObjects, completedObjects, incrementCompletedObjects, completedAt)
		if !isValid {
			log.Println(errorMsg)
			return false, errorMsg
		} else {
			return true, ""
		}
	} else {
		appUpdateFunction := os.Getenv("APP_UPDATE_EXECUTION_FUNCTION")
		if appUpdateFunction == "" {
			errorMsg := "APP_UPDATE_EXECUTION_FUNCTION is not in the environment"
			log.Println(errorMsg)
			return false, errorMsg
		}

		var executionID string
		var executionType string

		if c.ScanExecutionID != "" {
			executionID = c.ScanExecutionID
			executionType = "scan"
		} else if c.SyncExecutionID != "" {
			executionID = c.SyncExecutionID
			executionType = "sync"
		} else {
			errorMsg := "Missing required field: either 'scanExecutionId' or 'syncExecutionId' must be provided"
			log.Println(errorMsg)
			return false, errorMsg
		}

		payload := map[string]interface{}{
			"type":        executionType,
			"executionId": executionID,
		}

		// Only include optional fields if they are provided (not nil)
		if status != nil {
			payload["status"] = *status
		}
		if totalObjects != nil {
			payload["totalObjects"] = *totalObjects
		}
		if completedObjects != nil {
			payload["completedObjects"] = *completedObjects
		}
		if incrementCompletedObjects != nil {
			payload["incrementCompletedObjects"] = *incrementCompletedObjects
		}
		if completedAt != nil {
			payload["completedAt"] = *completedAt
		}

		payloadJSON, err := json.Marshal(payload)
		if err != nil {
			errorMsg := fmt.Sprintf("Error marshaling payload: %v", err)
			log.Println(errorMsg)
			return false, errorMsg
		}

		url := fmt.Sprintf("%s/async-function/%s", os.Getenv("OPENFAAS_GATEWAY"), appUpdateFunction)
		req, err := http.NewRequest("POST", url, bytes.NewBuffer(payloadJSON))
		if err != nil {
			errorMsg := fmt.Sprintf("Error creating request: %v", err)
			log.Println(errorMsg)
			return false, errorMsg
		}

		req.Header.Set("Content-Type", "application/json")
		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()
		req = req.WithContext(ctx)

		client := &http.Client{}
		resp, err := client.Do(req)
		if err != nil {
			errorMsg := fmt.Sprintf("Error: %v", err)
			log.Println(errorMsg)
			return false, errorMsg
		}
		defer resp.Body.Close()

		if resp.StatusCode == 202 {
			return true, ""
		} else {
			body, _ := io.ReadAll(resp.Body)
			errorMsg := fmt.Sprintf("Status %d: %s", resp.StatusCode, string(body))
			log.Println(errorMsg)
			return false, errorMsg
		}
	}
}

func getSecrets(localRun bool) map[string]string {
	secrets := make(map[string]string)
	secretsDir := "/var/openfaas/secrets/"

	entries, err := os.ReadDir(secretsDir)
	if err != nil {
		log.Printf("Error reading secrets directory: %v\n", err)
		return secrets
	}

	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}

		filename := entry.Name()
		secretPath := filepath.Join(secretsDir, filename)

		var keyName string
		if localRun {
			// For local run, use the filename as-is (no scan ID removal)
			keyName = filename
		} else {
			// For non-local run, remove the last 9 characters (dash + 8 chars scan ID)
			if len(filename) > 9 && filename[len(filename)-9:len(filename)-8] == "-" {
				keyName = filename[:len(filename)-9] // Remove last 9 characters (-abcd1234)
			} else {
				log.Printf("Skipping secret file with unexpected format: %s\n", filename)
				continue
			}
		}

		// Convert dash-separated to camelCase
		keyParts := strings.Split(keyName, "-")
		var camelKey string
		if len(keyParts) > 1 {
			// First part stays lowercase, subsequent parts are capitalized
			camelKey = keyParts[0]
			for _, part := range keyParts[1:] {
				camelKey += strings.Title(part)
			}
		} else {
			camelKey = keyParts[0]
		}

		// Read the secret content
		content, err := os.ReadFile(secretPath)
		if err != nil {
			log.Printf("Error reading secret file %s: %v\n", filename, err)
		} else {
			secrets[camelKey] = strings.TrimSpace(string(content))
			log.Printf("Loaded secret: %s\n", camelKey)
		}
	}

	return secrets
}

func Handle(w http.ResponseWriter, r *http.Request) {
	event := Event{
		Body:    nil,
		Headers: r.Header,
		Method:  r.Method,
		Query:   r.URL.Query(),
		Path:    r.URL.Path,
	}

	if r.Body != nil {
		defer r.Body.Close()
		body, _ := io.ReadAll(r.Body)
		event.Body = body
	}

	hostname := os.Getenv("HOSTNAME")
	if hostname == "" {
		hostname = "localhost"
	}

	context := Context{
		Hostname:     hostname,
		ScanID:       os.Getenv("SCAN_ID"),
		SyncID:       os.Getenv("SYNC_ID"),
		RunLocal:     os.Getenv("RUN_LOCAL"),
		FunctionType: os.Getenv("FUNCTION_TYPE"),
	}

	if context.RunLocal == "" {
		context.RunLocal = "false"
	}

	// Parse CONFIG
	configStr := os.Getenv("CONFIG")
	if configStr != "" {
		err := json.Unmarshal([]byte(configStr), &context.Config)
		if err != nil {
			http.Error(w, fmt.Sprintf("Invalid JSON in CONFIG: %v", err), http.StatusBadRequest)
			return
		}
	}

	localRun := context.RunLocal == "true"

	if localRun {
		if context.Config == nil {
			http.Error(w, "CONFIG is required when RUN_LOCAL is true", http.StatusBadRequest)
			return
		}

		if context.FunctionType == "" {
			http.Error(w, "FUNCTION_TYPE is required when RUN_LOCAL is true", http.StatusBadRequest)
			return
		}

		// Validate request body against config
		var requestData map[string]interface{}
		if len(event.Body) > 0 {
			err := json.Unmarshal(event.Body, &requestData)
			if err != nil {
				http.Error(w, fmt.Sprintf("Invalid JSON in request body: %v", err), http.StatusBadRequest)
				return
			}

			isValid, errorMsg := ValidateRequestSchema(context.Config, requestData, context.FunctionType)
			if !isValid {
				http.Error(w, errorMsg, http.StatusBadRequest)
				return
			}
		}
	}

	// Load secrets from OpenFaaS secret files
	context.Secrets = getSecrets(localRun)

	// Parse execution IDs from request body
	if len(event.Body) > 0 {
		var requestData map[string]interface{}
		json.Unmarshal(event.Body, &requestData)
		if scanExecutionID, ok := requestData["scanExecutionId"].(string); ok {
			context.ScanExecutionID = scanExecutionID
		}
		if syncExecutionID, ok := requestData["syncExecutionId"].(string); ok {
			context.SyncExecutionID = syncExecutionID
		}
	}

	if len(context.Secrets) == 0 {
		log.Println("Warning: No secrets loaded from secret files")
	} else {
		log.Printf("Loaded %d secrets from secret files\n", len(context.Secrets))
	}

	startedAt := time.Now().UTC().Format(time.RFC3339)
	responseData := handleFunction(event, context)
	completedAt := time.Now().UTC().Format(time.RFC3339)

	responseJSON, _ := json.Marshal(responseData)
	log.Printf("Response data: %s\n", string(responseJSON))

	// Add timestamps for successful responses
	if context.FunctionType == "test-connection" && responseData.StatusCode == 200 {
		if responseData.Body == nil {
			responseData.Body = make(map[string]interface{})
		}
		responseData.Body["startedAt"] = startedAt
		responseData.Body["completedAt"] = completedAt
	} else if context.FunctionType == "access-scan" && responseData.StatusCode == 200 {
		if responseData.Body == nil {
			responseData.Body = make(map[string]interface{})
		}
		responseData.Body["startedAt"] = startedAt
		responseData.Body["completedAt"] = completedAt
	}

	if localRun {
		isValid, errorMsg := ValidateResponse(context.FunctionType, responseData)
		if !isValid {
			responseData = context.ErrorResponse(false, errorMsg)
		}
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(responseData.StatusCode)

	if responseData.Body != nil {
		json.NewEncoder(w).Encode(responseData.Body)
	}
}

// handleFunction is a placeholder for the actual function implementation
func handleFunction(event Event, context Context) Response {
	// This would call the actual handler implementation
	// For now, return a simple response
	return Response{
		StatusCode: 200,
		Body:       map[string]interface{}{"message": "Hello from Netwrix Go handler"},
	}
}